platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,720 @@
+missing parent is received, the child may be emitted. The current implementation uses what is known as the "orphan
+buffer" for this purpose.
+
+Since Event Intake will also maintain some buffers, it needs to know about the progression of the hashgraph,
+so it can evict old events. In this case, the "orphan buffer" holds events until either the parent events have
+arrived, or the events have become ancient due to the advancement of the "non-ancient event window" and the event is
+dropped from the buffer. This document does not prescribe the existence of the orphan buffer or the method by which
+events are sorted and emitted in topological order, but it does describe a method by which old events can be dropped.
+
+#### Birth-Round Filtering
+
+When an event is created, it is assigned a "birth round". This is the most recent consensus round on the node at the
+time the event is created. Two different nodes in the same network might be working on very different moments in the
+hashgraph timeline. One node, Alice, may be working on round 100 while a better-connected or faster node, Bob is working
+on round 200. When Alice creates an event, it will be for birth-round 100, while an event created by Bob at the same
+instant would be for birth-round 200.
+
+In a sense, while both events are created at the same wall-clock time, Alice's event is actually older, because of what
+Alice understood about the state of the world at the time the event was created.
+
+It is not possible for any one node to get much farther ahead of all other nodes, since the only way to have a newer
+birth-round is to advance the hashgraph, and that requires 2/3 of the network by stake weight. Therefore, in this
+example, Alice is not just 100 rounds behind Bob, she must be 100 rounds behind at least 2/3 of the network by
+stake weight, or, Bob is lying. He may create events at round 200, but not actually have a hashgraph that has advanced
+to that round.
+
+Each event must be validated using the roster associated with its birth-round. If Alice is far behind Bob, and she
+receives an event for a birth-round she doesn't have the roster for, then she cannot validate the event. So the Event
+Intake module will drop events with a birth-round too far in the future.
+
+If the event has been dropped, it is possible to fetch missing events from neighbors using Gossip or, if the node has
+fallen so far behind it can no longer fetch events from neighbors in gossip, then the node will reconnect.
+
+#### Self Events
+
+Events are not only given to the Event Intake system through gossip. Self events (those events created by the node
+itself) are also fed to Event Intake. These events **may** bypass some steps in the pipeline. For example, self-events
+do not need validation. Likewise, when replaying events from the pre-consensus event buffer, those checks are not needed
+(since they have already been proved valid and are in topological order).
```

Comment:
This seems to presume that disk storage never fails and always returns the exact data written to it previously. This may not be true.

Reply:
In this context, disk failure can be considered to be in the realm of byzantine failures. That is, we lump disk failure into the group of < 1/3 nodes that we tolerate failures in when reasoning about the security of the system as a whole. For example, what if a bits flip inside the compiled jar file (also stored on disk) which cause the program to behave in arbitrarily unpredictable ways?

In summary, I don't think it's necessary to  add caveats about possible disk failure. If a disk fails, the node is no longer an "honest" node in the general sense.